core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1448,58 +1512,73 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
           }
 
           // now do all the work: invites, additions, removals, notifications
-          for {
-            maybeWorkspace <- dataSource.inTransaction { dataAccess =>
-              dataAccess.workspaceQuery.findByName(workspaceName)
+          inviteNotifications <- Future.traverse(userToInvite) { invite =>
+            samDAO.inviteUser(invite, ctx.userInfo).map { _ =>
+              Notifications.WorkspaceInvitedNotification(
+                WorkbenchEmail(invite),
+                WorkbenchUserId(ctx.userInfo.userSubjectId.value),
+                NotificationWorkspaceName(workspaceName.namespace, workspaceName.name),
+                workspace.bucketName
+              )
             }
-            workspace = maybeWorkspace.getOrElse(throw new RawlsException(s"workspace $workspaceName not found"))
-
-            inviteNotifications <- Future.traverse(userToInvite) { invite =>
-              samDAO.inviteUser(invite, ctx.userInfo).map { _ =>
-                Notifications.WorkspaceInvitedNotification(
-                  WorkbenchEmail(invite),
-                  WorkbenchUserId(ctx.userInfo.userSubjectId.value),
-                  NotificationWorkspaceName(workspaceName.namespace, workspaceName.name),
-                  workspace.bucketName
+          }
+
+          // do additions before removals so users are not left unable to access the workspace in case of errors that
+          // lead to incomplete application of these changes, remember: this is not transactional
+          _ <- Future.traverse(policyAdditions) { case (policyName, email) =>
+            workspace.workspaceType match {
+              case WorkspaceType.RawlsWorkspace =>
+                samDAO.addUserToPolicy(SamResourceTypeNames.workspace,
+                                       workspace.workspaceId,
+                                       policyName,
+                                       email,
+                                       ctx.userInfo
+                )
+              case WorkspaceType.McWorkspace =>
+                Future(
+                  Option(IamRole.fromValue(policyName.value.toUpperCase)).map { role =>
```

Comment:
🤔 This will mask the case of IamRole giving back null here won't it? We should throw an exception in that case, or at least log an error.